lib/Sema/CSDiag.cpp

```diff
@@ -1060,19 +1062,38 @@ CalleeCandidateInfo::evaluateCloseness(Type candArgListType,
             // Multiple archetypes, too complicated.
             return { CC_ArgumentMismatch, {}};
           if (rArgType->isEqual(matchingArgType)) {
-            continue;
+            if (nonSubstitutableArgs == 0) {
+              continue;
+            } else {
+              ++nonSubstitutableArgs;
+              // Fallthrough, this is nonsubstitutable, so mismatches as well.
+            }
           } else {
-            paramType = matchingArgType;
-            // Fallthrough as mismatched arg, comparing nearness to archetype
-            // bound type.
+            if (nonSubstitutableArgs == 0) {
+              paramType = matchingArgType;
+              // Fallthrough as mismatched arg, comparing nearness to archetype
+              // bound type.
+            } else if (nonSubstitutableArgs == 1) {
+              // If we have only one nonSubstitutableArg so far, then this different
+              // type might be the one that we should be substituting for instead.
+              // Note that failureInfo is already set correctly for that case.
+              auto archetype = paramType->getAs<ArchetypeType>();
```

Comment:
This is a minor issue with the existing code, but please use ->castTo( instead of getAs, since it is statically knowable that you have an archetype here.  Otherwise, I'd expect to see null handling code.